File: code/ch12/java/RefactoringMoves.java
// §12.6.3 Refactoring Under Green Tests — two moves on one function:
// replace magic number with named constant, then introduce guard clauses.
// The same suite runs green before, between, and after the moves.
// Run: java -ea RefactoringMoves.java
public class RefactoringMoves {
  record Patient(String id) {}
  record Slot(boolean open) {}

  interface Booker { boolean canBook(Patient patient, Slot slot, int bookedToday); }

  // before: magic number, three-deep nesting
  static boolean canBook(Patient patient, Slot slot, int bookedToday) {
    if (patient != null) {
      if (slot.open()) {
        if (bookedToday < 8) {
          return true;
        } else {
          return false;
        }
      } else {
        return false;
      }
    } else {
      return false;
    }
  }

  // move 1: replace magic number with named constant
  static final int MAX_DAILY_BOOKINGS = 8;

  static boolean canBookM1(Patient patient, Slot slot, int bookedToday) {
    if (patient != null) {
      if (slot.open()) {
        if (bookedToday < MAX_DAILY_BOOKINGS) {
          return true;
        } else {
          return false;
        }
      } else {
        return false;
      }
    } else {
      return false;
    }
  }

  // move 2: introduce guard clauses
  static boolean canBookM2(Patient patient, Slot slot, int bookedToday) {
    if (patient == null) return false;
    if (!slot.open()) return false;
    return bookedToday < MAX_DAILY_BOOKINGS;
  }

  static void runSuite(Booker fn) {
    Patient p = new Patient("p1");
    assert fn.canBook(p, new Slot(true), 0);
    assert fn.canBook(p, new Slot(true), 7);
    assert !fn.canBook(p, new Slot(true), 8);
    assert !fn.canBook(p, new Slot(false), 0);
    assert !fn.canBook(null, new Slot(true), 0);
  }

  public static void main(String[] args) {
    String[] names = {"before", "after move 1", "after move 2"};
    Booker[] fns = {RefactoringMoves::canBook, RefactoringMoves::canBookM1,
        RefactoringMoves::canBookM2};
    for (int i = 0; i < fns.length; i++) {
      runSuite(fns[i]);
      System.out.println(names[i] + ": 5 tests green");
    }
    for (Patient patient : new Patient[]{new Patient("p1"), null}) {
      for (Slot slot : new Slot[]{new Slot(true), new Slot(false)}) {
        for (int booked = 0; booked < 12; booked++) {
          assert canBook(patient, slot, booked) == canBookM2(patient, slot, booked);
        }
      }
    }
    System.out.println("before/after agree on all 48 input combinations");
  }
}

File: code/ch12/javascript/refactoring_moves.js
// §12.6.3 Refactoring Under Green Tests — two moves on one function:
// replace magic number with named constant, then introduce guard clauses.
// The same suite runs green before, between, and after the moves.
// Run: node refactoring_moves.js
const assert = require("node:assert/strict");

// before: magic number, three-deep nesting
function canBook(patient, slot, bookedToday) {
  if (patient !== null) {
    if (slot.open) {
      if (bookedToday < 8) {
        return true;
      } else {
        return false;
      }
    } else {
      return false;
    }
  } else {
    return false;
  }
}

// move 1: replace magic number with named constant
const MAX_DAILY_BOOKINGS = 8;

function canBookM1(patient, slot, bookedToday) {
  if (patient !== null) {
    if (slot.open) {
      if (bookedToday < MAX_DAILY_BOOKINGS) {
        return true;
      } else {
        return false;
      }
    } else {
      return false;
    }
  } else {
    return false;
  }
}

// move 2: introduce guard clauses
function canBookM2(patient, slot, bookedToday) {
  if (patient === null) return false;
  if (!slot.open) return false;
  return bookedToday < MAX_DAILY_BOOKINGS;
}

function runSuite(fn) {
  assert.equal(fn("p1", { open: true }, 0), true);
  assert.equal(fn("p1", { open: true }, 7), true);
  assert.equal(fn("p1", { open: true }, 8), false);
  assert.equal(fn("p1", { open: false }, 0), false);
  assert.equal(fn(null, { open: true }, 0), false);
}

for (const [name, fn] of [["before", canBook], ["after move 1", canBookM1],
  ["after move 2", canBookM2]]) {
  runSuite(fn);
  console.log(`${name}: 5 tests green`);
}
for (const patient of ["p1", null]) {
  for (const slot of [{ open: true }, { open: false }]) {
    for (let booked = 0; booked < 12; booked++) {
      assert.equal(canBook(patient, slot, booked), canBookM2(patient, slot, booked));
    }
  }
}
console.log("before/after agree on all 48 input combinations");
